src/groth16/prover.rs

```diff
@@ -293,150 +317,245 @@ where
+                    b.coset_fft(&worker, &mut fft_kern)?;
+
+                    info!("c: ifft");
+                    c.ifft(&worker, &mut fft_kern)?;
+                    c.coset_fft(&worker, &mut fft_kern)?;
+
+                    info!("fft collect");
+
+                    a.mul_assign(&worker, &b);
+                    a.sub_assign(&worker, &c);
+
+                    drop(b);
+                    drop(c);
+
+                    info!("coset");
+                    a.divide_by_z_on_coset(&worker);
+                    a.icoset_fft(&worker, &mut fft_kern)?;
+
+                    let a = a.into_coeffs();
+                    let a_len = a.len() - 1;
+
+                    Ok(Arc::new(
+                        a.into_iter()
+                         .take(a_len)
+                         .map(|s| s.0.into_repr())
+                         .collect::<Vec<_>>(),
+                    ))
+                })
+                .collect::<Result<Vec<_>, SynthesisError>>()
+        },
+        let params_h = {
+            params.get_h(n)
+        },
+        let params_l = {
+            params.get_l(provers_len)
+        },
+        let a_source = {
+            params.get_a(input_len, a_aux_density_total)
+        },
+        let b_g1_source = {
+            params.get_b_g1(b_input_density_total, b_aux_density_total)
+        },
+        let b_g2_source = {
+            params.get_b_g2(b_input_density_total, b_aux_density_total)
+        }
+    };
 
-    drop(fft_kern);
+    let a_s = a_s?;
+    let params_h = params_h?;
+    let params_l = params_l?;
+    let (a_inputs_source, a_aux_source) = a_source?;
+    let (b_g1_inputs_source, b_g1_aux_source) = b_g1_source?;
+    let (b_g2_inputs_source, b_g2_aux_source) = b_g2_source?;
+
+    info!("fft done");
     let mut multiexp_kern = Some(LockedMultiexpKernel::<E>::new(log_d, priority));
 
-    let h_s = a_s
-        .into_iter()
-        .map(|a| {
-            let h = multiexp(
-                &worker,
-                params.get_h(a.len())?,
-                FullDensity,
-                a,
-                &mut multiexp_kern,
-            );
-            Ok(h)
-        })
-        .collect::<Result<Vec<_>, SynthesisError>>()?;
+    let params_h_ref = &params_h;
+    let worker = &worker;
+    let mkern = &mut multiexp_kern;
+    let provers_ref = &provers;
+    par! {
+        let h_s = {
+            info!("h_s");
+            a_s.into_iter()
+                .map(|a| {
+                    let h = multiexp(
+                        worker,
+                        params_h_ref.clone(),
+                        FullDensity,
+                        a,
+                        mkern,
+                    );
+                    Ok(h)
+                })
+                .collect::<Result<Vec<_>, SynthesisError>>()
+        },
+        let input_assignments = {
+            info!("input_assignments");
+            provers_ref
+                .par_iter()
```

Comment:
It looks like the `par!` blocks are outside of threadpool calls, which means that this par_iter will be unconstrained.  Is that correct?

Reply:
fixed